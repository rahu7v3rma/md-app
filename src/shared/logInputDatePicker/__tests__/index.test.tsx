import moment from 'moment';
import React from 'react';
import renderer, { act } from 'react-test-renderer';

import LogInputDatePicker from '@/shared/logInputDatePicker';

function checkIfCalendarIsClosed(tree: renderer.ReactTestRenderer) {
    let calendarContainer;
    try {
        calendarContainer = tree.root.findByProps({
            testID: 'calendarContainer'
        });
    } catch (error) {}
    expect(calendarContainer).toBeUndefined();
}

it('logInputDatePicker snapshot', async () => {
    const tree = renderer.create(
        <LogInputDatePicker
            selectedDate={'2023-01-01'}
            onDateSelected={(_: Date) => _}
        />
    );
    expect(tree).toMatchSnapshot();
});

it('logInputDatePicker renders without crashing', async () => {
    const tree = renderer.create(
        <LogInputDatePicker
            selectedDate={moment().format('YYYY-MM-DD')}
            onDateSelected={(_: Date) => _}
        />
    );
    expect(tree.root).toBeTruthy();
});

it('logInputDatePicker calendar is initially hidden', async () => {
    const tree = renderer.create(
        <LogInputDatePicker
            selectedDate={moment().format('YYYY-MM-DD')}
            onDateSelected={(_: Date) => _}
        />
    );

    checkIfCalendarIsClosed(tree);
});

it('logInputDatePicker clicking LogInput opens the calendar', async () => {
    const tree = renderer.create(
        <LogInputDatePicker
            selectedDate={moment().format('YYYY-MM-DD')}
            onDateSelected={(_: Date) => _}
        />
    );
    const logInputContainer = tree.root.findByProps({
        testID: 'logInputContainer'
    }).props;
    await act(() => logInputContainer.onPress());

    const calendarContainer = tree.root.findByProps({
        testID: 'calendarContainer'
    }).props;

    expect(calendarContainer).toBeTruthy();
});

it('logInputDatePicker clicking Today button updates the selected date', async () => {
    const selectedDate = moment().format('YYYY-MM-DD');
    const onDateSelected = jest.fn().mockImplementation((date: Date) => {
        expect(moment(date).format('YYYY-MM-DD')).toEqual(selectedDate);
    });

    const tree = renderer.create(
        <LogInputDatePicker
            selectedDate={selectedDate}
            onDateSelected={onDateSelected}
        />
    );
    const logInputContainer = tree.root.findByProps({
        testID: 'logInputContainer'
    }).props;
    await act(() => logInputContainer.onPress());

    const todayButton = tree.root.findByProps({
        testID: 'todayButton'
    }).props;
    await act(() => todayButton.onPress());
    expect(onDateSelected).toHaveBeenCalledTimes(1);
});

it('logInputDatePicker selecting a date closes the calendar and updates the selected date', async () => {
    const selectedDateString = '2023-01-01';
    const onDateSelected = jest.fn().mockImplementation((date: Date) => {
        expect(moment(date).format('YYYY-MM-DD')).toEqual(selectedDateString);
    });

    const onCalendarSheetClosed = jest.fn();

    const tree = renderer.create(
        <LogInputDatePicker
            selectedDate={moment().format('YYYY-MM-DD')}
            onDateSelected={onDateSelected}
            onCalendarClosed={onCalendarSheetClosed}
        />
    );
    const logInputContainer = tree.root.findByProps({
        testID: 'logInputContainer'
    }).props;
    await act(() => logInputContainer.onPress());

    const calendar = tree.root.findByProps({
        testID: 'calendar'
    }).props;
    await act(() => calendar.onDayPress({ dateString: selectedDateString }));
    expect(onDateSelected).toHaveBeenCalledTimes(1);
    expect(onCalendarSheetClosed).toHaveBeenCalledTimes(1);
});

it('logInputDatePicker initializes with correct initial state', async () => {
    const initialSelectedDate = moment().format('YYYY-MM-DD');
    const disabled = true;
    const minDate = moment().subtract(1, 'year').format('YYYY-MM-DD');

    const tree = renderer.create(
        <LogInputDatePicker
            selectedDate={initialSelectedDate}
            disabled={disabled}
            minDate={minDate}
            onDateSelected={(_: Date) => _}
        />
    );

    // Open calendar
    const logInputContainer = tree.root.findByProps({
        testID: 'logInputContainer'
    }).props;
    await act(() => logInputContainer.onPress());

    if (!disabled) {
        const calendar = tree.root.findByProps({
            testID: 'calendar'
        }).props;

        expect(calendar.minDate).toBe(minDate);
        expect(calendar.current).toBe(initialSelectedDate);
    } else {
        // check if calendar is still closed after clicking LogInput when disabled
        checkIfCalendarIsClosed(tree);
    }
});
