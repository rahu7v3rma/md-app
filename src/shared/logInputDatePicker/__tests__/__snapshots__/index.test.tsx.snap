// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logInputDatePicker snapshot 1`] = `
[
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 16,
        "elevation": 5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 8,
        "opacity": 1,
        "paddingHorizontal": 20,
        "paddingVertical": 18,
        "shadowColor": "rgba(0,0,0,0.5)",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 16,
      }
    }
    testID="logInputContainer"
  >
    <Text
      adjustsFontSizeToFit={true}
      style={
        [
          {
            "fontFamily": "Poppins",
          },
          {
            "flex": 1,
            "lineHeight": 20,
          },
          {
            "color": "#58595A",
            "fontSize": 14,
            "fontWeight": "500",
            "lineHeight": 22,
          },
        ]
      }
      testID="fieldNameTxt"
    >
      Date
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <Text
        adjustsFontSizeToFit={true}
        style={
          [
            {
              "fontFamily": "Poppins",
            },
            {
              "flex": 1,
              "lineHeight": 20,
              "paddingRight": 10,
              "textAlign": "right",
            },
            {
              "color": "#271A51",
              "fontSize": 14,
              "fontWeight": "600",
              "lineHeight": 22,
            },
          ]
        }
        testID="valueTxt"
      >
        01 Jan 2023
      </Text>
    </View>
  </View>,
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    supportedOrientations={
      [
        "portrait",
        "portrait-upside-down",
        "landscape",
        "landscape-left",
        "landscape-right",
      ]
    }
    transparent={true}
    visible={false}
  />,
]
`;
